File: advent-of-code/bin/day03/day-03.py
#!/usr/bin/env python

class Houses:
	all = {}

	def deliver( self, x, y ):
		key = str(x) + '::' + str(y)
		self.all[key] = 1 + ( self.all[key] if (key in self.all) else 0 )

	def numHouses( self ):
		total = 0;
		for key in self.all:
			total += 1
		return total

	def __init__( self ):
		self.all = {}

class Santa:
	x = 0
	y = 0
	houses = None

	def deliver( self ):
		self.houses.deliver( self.x, self.y )

	def move( self, dir ):
		if   ( dir == '<' ): self.x -= 1
		elif ( dir == '>' ): self.x += 1
		elif ( dir == '^' ): self.y -= 1
		elif ( dir == 'v' ): self.y += 1

	def report( self ):
		return "pos(" + str(self.x) + "," + str(self.y) + ")"

	def __init__( self, houses ):
		self.x = 0
		self.y = 0
		self.houses = houses

def thisYear( instructions ):
	houses = Houses()
	print "Houses visited:", houses.numHouses()

	santa = Santa( houses )
	santa.deliver()
	for dir in instructions:
		santa.move( dir )
		santa.deliver()
	return houses.numHouses()

def nextYear( instructions ):
	houses = Houses()
	print "Houses visited:", houses.numHouses()

	santa = Santa( houses )
	santa.deliver()

	roboSanta = Santa( houses )
	roboSanta.deliver()

	current = santa
	for dir in instructions:
		print "\n"
		print "Instruction: ", dir
		current.move( dir )
		current.deliver()
		current = roboSanta if ( current == santa ) else santa
		print "Santa is at", santa.report()
		print "RoboSanta is at", roboSanta.report()
		print "Houses visited:", houses.numHouses()
	
	return houses.numHouses()

with open( 'day-03.txt', 'r' ) as input_file:
	data = input_file.read()

print "In the first year, Santa delivers to ", thisYear( data )
print "In the second year, Santa and Robo-Santa delivers to ", nextYear( data ) 
	
	
			

		

File: advent-of-code/src/day03/Day03.java
package day03;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Hashtable;

class Neighborhood {
	private Hashtable<String, Integer> homes;

	public Neighborhood() {
		homes = new Hashtable<String, Integer>();
	}

	public void deliver(int x, int y) {
		String key = x + "::" + y;
		Integer current = homes.getOrDefault(key, 0);
		homes.put(key, current + 1);
	}

	public int homesVisited() {
		return homes.size();
	}
}

class JollyOldElf {
	private int x;
	private int y;
	private Neighborhood homes;

	public JollyOldElf(Neighborhood toVisit) {
		x = 0;
		y = 0;
		homes = toVisit;
	}

	public void deliver() {
		homes.deliver(x, y);
	}

	public void move(char dir) {
		if (dir == '^')
			y--;
		else if (dir == 'v')
			y++;
		else if (dir == '<')
			x--;
		else if (dir == '>')
			x++;
	}

	public String report() {
		return "pos(" + x + "," + y + ")";
	}

}

public class Day03 {

	public static int firstYear(String instructions) {
		Neighborhood hood = new Neighborhood();

		JollyOldElf santa = new JollyOldElf(hood);
		santa.deliver();

		for (char dir : instructions.toCharArray()) {
			santa.move(dir);
			santa.deliver();
		}

		return hood.homesVisited();
	}

	public static int secondYear(String instructions) {
		Neighborhood hood = new Neighborhood();

		JollyOldElf santa = new JollyOldElf(hood);
		santa.deliver();

		JollyOldElf roboSanta = new JollyOldElf(hood);
		roboSanta.deliver();

		JollyOldElf active = santa;
		for (char dir : instructions.toCharArray()) {
			active.move(dir);
			active.deliver();
			active = (active == santa ? roboSanta : santa);
		}

		return hood.homesVisited();
	}

	public static void main(String[] args) throws IOException {
		byte[] encoded = Files.readAllBytes(Paths.get("input", "day-03.txt"));
		String instructions = new String(encoded, StandardCharsets.UTF_8);

		System.out.println("In year one, Santa delivers to " + firstYear(instructions) + " homes");
		System.out.println("In year two, Santa and RoboSanta deliver to " + secondYear(instructions) + " homes");
	}
}
